Initial commit for strongly-typed CRDs

albert-zhong opened the pull request:
Here's my initial attempt at generating some strongly-typed CRDs. I'm not exactly sure which folder this code should go in, so I just made a new one called `crds/` for now. When time comes I will move it to a more appropriate location. This is my first time working with Go, so I would love feedback/criticism on my coding style and programming structure.

So far, this tool takes in the path for a resourcedefinition.yaml file and prints out the generated code to standard output. For example you can run:
```bash
$ go run main.go gen.go parse.go nestedMap.go <resourcedefinition.yaml> > <output.ts>
```

For the CronTab example, it should generate this:
```ts
import * as pulumi from "@pulumi/pulumi";
import * as k8s from "@pulumi/kubernetes";

export class CronTabDefinition extends k8s.apiextensions.v1.CustomResourceDefinition {
	constructor(name: string, opts?: pulumi.CustomResourceOptions) {
		super(name, {
			"apiVersion": "apiextensions.k8s.io/v1",
			"kind": "CustomResourceDefinition",
			"metadata": {
				"name": "crontabs.stable.example.com"
			},
			"spec": {
				"group": "stable.example.com",
				"names": {
					"kind": "CronTab",
					"plural": "crontabs",
					"shortNames": [
						"ct"
					],
					"singular": "crontab"
				},
				"scope": "Namespaced",
				"versions": [
					{
						"name": "v1",
						"schema": {
							"openAPIV3Schema": {
								"properties": {
									"spec": {
										"properties": {
											"cronSpec": {
												"type": "string"
											},
											"image": {
												"type": "string"
											},
											"replicas": {
												"type": "integer"
											}
										},
										"required": [
											"cronSpec"
										],
										"type": "object"
									}
								},
								"type": "object"
							}
						},
						"served": true,
						"storage": true
					}
				]
			}
		}, opts)
	}
}

export class CronTab extends k8s.apiextensions.CustomResource {
	constructor(name: string, args?: CronTabArgs, opts?: pulumi.CustomResourceOptions) {
		super(name, { apiVersion: "stable.example.com/v1", kind: "CronTab", ...args }, opts)
	}
}

export interface CronTabArgs {
	metadata?: pulumi.Input<k8s.types.input.meta.v1.ObjectMeta>;
	spec?: CronTabSpecArgs;
}

export interface CronTabSpecArgs {
	cronSpec: pulumi.Input<string>;
	image?: pulumi.Input<string>;
	replicas?: pulumi.Input<number>;
}
```

**Two main issues I ran into:**

1. Couldn't unmarshal CRD YAML files into a struct, so for now I used a hacky `map[interface{}]interface{}` way

One thing I did that probably isn't good practice is that I first unmarshalled the CRD YAML file into a `map[interface{}]interface{}`. I then created a wrapper struct called `NestedMap` that treated it as an `interface{}`, with methods to quickly cast and traverse down key/values. Although it works for the CronTab example in the Kubernetes documentation, this is definitely a very hacky solution, so I expect it will run into a lot of issues with types for other CRDs.

I chose that solution for now because I was having some trouble unmarshalling a CRD YAML file into an actual struct. At first, I thought I could just copy the already generated `customResourceDefinitionArgs` struct in the pulumi-kubernetes Go SDK, but I ran into some strange issues with the tags and fields. For example the `ApiVersion` string field could not get unmarshalled unless I changed the tag to `yaml:"apiVersion"`, but the `kind` field unmarshalled just fine even though the tag is `pulumi:"kind"`. I then tried using a package called [retag](https://github.com/sevlyar/retag) to programmatically change every tag from `pulumi` to `yaml`, but I couldn't quite get it to work. I'm going to work more on this issue tommorow so that I can hopefully stop using my `NestedMap` struct.

2. Wasn't able to effectively refractor existing codegen tools
As Justin suggested I first tried to expose and refractor the existing codegen tools, but I couldn't quite figure out how to get them to work. For now I just hand-rolled some functions to generate the interfaces/classes/types necessary, and as I understand more of the codebase I'll incremently use them instead. Fortunately since I made the generated resource class and resource definition class just extend CustomResource and CustomResourceDefinition, this wasn't a lot of work at all. All I had to do was generate method/class headers, calls to super, and that was it.

**Goals for the future:**
- So far this only works for NodeJS, but eventually it needs to work for Python, C#, and Go as well
- I need to understand the "contours" of the CRD Schema better to implement more strong-typing features. For now my code just gets the name and type of each entry in `.spec.versions[0].schema.openAPIV3Schema.properties.spec.properties`, and then makes them required if they appear in `.spec.versions[0].schema.openAPIV3Schema.properties.required`. However, there's a lot of other fields/features in the OpenAPI schema that I could probably represent in a generated code snippet, so I'll take some time over this summer to figure those out.

lblackstone commented on crds/README.md:
````diff
@@ -0,0 +1,119 @@
+schema strongly typed, you can run:
+
+```bash
+$ go run main.go gen.go parse.go nestedMap.go <resourcedefinition.yaml> > <output.ts>
````
```suggestion
$ go run . <resourcedefinition.yaml>
```

lblackstone commented on crds/README.md:
````diff
@@ -0,0 +1,119 @@
+
+However with this tool, we can simply can simply run:
+```bash
+$ go run main.go gen.go parse.go nestedMap.go resourcedefinition.yaml > crontab.ts
````
```suggestion
$ go run . resourcedefinition.yaml > crontab.ts
```

lblackstone commented on crds/parse.go:
```diff
@@ -0,0 +1,96 @@
+
+// NewResourceDefinition returns a ResourceDefinition by parsing out all
+// required information from a NestedMap.
+func NewResourceDefinition(yamlFile []byte) ResourceDefinition {
```
I'd recommend taking a look at this code: https://github.com/pulumi/pulumi-kubernetes/blob/b493881669be46a881a0b4ca96d6d5656bb42d76/provider/pkg/provider/invoke_decode_yaml.go#L32-L41

This method uses the Kubernetes packages to decode the YAML into the `Unstructured` type. This type contains some convenience methods for working with k8s resources, but mostly relies on `map[string]interface{}`. Unlike the built in k8s resources, we don't have types defined for CustomResources, so we're stuck with the generic map type.

lblackstone commented on crds/gen.go:
```diff
@@ -0,0 +1,73 @@
+	return fmt.Sprintf("%s.%s", kubernetesName, modulePath)
+}
+
+func (r ResourceDefinition) GenerateNodeJS(w io.Writer) {
```
This is a good start for learning, but you'll eventually want to generate a Pulumi schema internally so that you can use our codegen packages.

[Here's the schema](https://github.com/pulumi/pulumi-random/blob/master/provider/cmd/pulumi-resource-random/schema.json) for a simple provider to give you an idea of what I'm talking about.

We have types defined already for this schema, so you can translate from the `map[string]interface{}` you decode from the YAML and convert directly to the schema types.

Schema types are defined [here](https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/schema.go)
Example usage is [here](https://github.com/pulumi/pulumi-kubernetes/blob/master/provider/pkg/gen/schema.go)

albert-zhong commented:
In my recent commit I made a few major changes:

- Switched from using my `NestedMap` struct to the `unstructured` package to unmarshal YAML
- Removed all hand-rolled code generation, replaced it with Pulumi schemas

Like before, you can still run `go run . <resourcedef.yaml> > <resource.ts>` to see the output.

As for now, only the strongly typed interfaces for CustomResourceArgs gets generated. It correctly generates interfaces for any arbitrarily nested `resourcedefinition.yaml` file, with proper `required`, `additionalProperties`, and `description` fields also filled.

So far all I do is recursively fill in the `types := make(map[string]pschema.ObjectTypeSpec)` map with as much information that could mapped over from the YAML file, and then I let Justin's example codegen handle the rest. I mainly relied on looking through the fields of `ObjectTypeSpec`, `PropertySpec`, and `TypeSpec` to see which things I could convert from a YAML file. 

Current issues/things to implement:

- [ ] I assume that we would want to represent an OpenAPI v3 Schema of `type: object`, empty `properties`, but `additionalProperties: type: string` as a `[key: string]: pulumi.Input<string>`. So far the codegen just creates an empty interface if `properties` is empty.
- [ ] Fix the input/output import
- [ ] Fix array types. For example, if we had some schema like:
```yaml
envSecrets:
   type: array
   items:
      type: string
```
Then I generate a `pschema.TypeSpec` that looks like:
```golang
_ := pschema.ObjectTypeSpec{
    Type: "array",
    Items: &pschema.TypeSpec{
        Type: &pschema.TypeSpec{
            Type: "string",
        },
    },
}
```
Which I want to generate `pulumi.Input<string[]>` in TypeScript. However as of now it incorrectly generates `pulumi.Input<pulumi.Input<string>[]>`.
- [ ] Add support for parsing `anyOf` within a schema, assuming that it directly corresponds to a type

Long-term goal:
As of now, these strongly typed interfaces just serve as a wrapper over the underlying `k8s.apiextensions.CustomResource` args input. Eventually, it would interesting if I could generate code to handle outputs as well.

justinvp commented on crds/main.go:
```diff
@@ -0,0 +1,266 @@
+	yamlPath := os.Args[1]
+	yamlFile, err := ioutil.ReadFile(yamlPath)
+	if err != nil {
+		log.Printf("yamlFile.Get err	#%v ", err)
```
Nit:

Once we move to `cobra`, it's slightly nicer as you can [just return the error](https://github.com/pulumi/tf2pulumi/blob/1e7538fb9a348411e030c08ea609722525c5f151/main.go#L58), but in the meantime, we should do something like:

```go
if err != nil {
	fmt.Fprintf(os.Stderr, "error: %v\n", err)
	os.Exit(-1)
}
```

Or if you want to provide more details, you could import `"github.com/pkg/errors"` and use `errors.Wrap`:

```go
if err != nil {
	fmt.Fprintf(os.Stderr, "error: %v\n", errors.Wrap(err, "reading the yaml file"))
	os.Exit(-1)
}
```

justinvp commented on crds/main.go:
```diff
@@ -0,0 +1,266 @@
+}
+
+func generateTypeScriptTypes(types map[string]schema.ObjectTypeSpec) (string, error) {
+	// Create a fake `PackageSpec` that includes the specifed types.
+	// This lets us call the public `nodejs.GeneratePackage` function, which generates code
+	// for an entire package (which isn't really what we want to do, but is existing code
+	// we have with a public entry point, that I'll leverage here for demonstration purposes).
+	// We then return contents of a certain in-memory file that is generated.
+
+	// Note: This is unlikely how we actually want to do the real code gen for the CRD types.
+	// See my comments at the bottom of this function.
+
+	// Create some a property map that references our types, that we can add
+	// as properties to a fake resource, otherwise, our types won't be generated
+	// by the Node.js codegen in `types/input.ts` or `types/output.ts` in-memory files.
```
Feel free to pare down the comments that I had written in my example throughout this file.

justinvp commented on crds/parse.go:
```diff
@@ -0,0 +1,96 @@
+	var m map[interface{}]interface{}
+	err := yaml.Unmarshal(yamlFile, &m)
+	if err != nil {
+		log.Fatalf("Unmarshal: %v", err)
```
Instead of logging the error, return it. Applies throughout.

albert-zhong commented:
As Justin suggested, I added the copyright header and wrote `os.Exit(-1)` where there were errors. I also implemented the parsing of `anyOf`, `allOf`, `oneOf`, with these approaches:

- `oneOf` was pretty simple to implement, since `pschema.TypeSpec` struct already has a field for `OneOf` sub-schemas. I just recursively generated TypeSpecs for sub-schemas and added them to the `OneOf` struct field.

- `allOf` was a bit more tricky, since there is no built-in field for that in `pschema.TypeSpec`. I wrote a function to pool together the properties, required properties, and descriptions of sub-schemas into a single schema.

- `anyOf` is not feasible to perfectly implement. If we have a type which satisfies 1+ schemas out of _n_ number of sub-schemas, then we would need a `oneOf` for _2^(n-1)_ schemas to capture all combinations. I settled for generating a single schema that was `allOf`, but setting all properties to be optional. This way users should be able to see the names of properties and their types for a schema, but unfortunately won't be able to tell if they actually satisfy 1+ of those schemas.

- No way to implement `not` unfortunately, so I just return an `any` type if the tool ever runs into them.

Things to still accomplish:

- Parse `discriminator` keywords
- Parse `$ref` values
- Use the `cobra` package for handling command line arguments
- Write unit tests for functions I wrote
- `main.go` is getting longer now, I'll probably move a lot of the codegen to a separate `gen.go` file
- Generate constructors for actually creating the custom resources, since right now I just create the args interfaces

lblackstone commented on crds/main.go:
```diff
@@ -0,0 +1,374 @@
+	return types
+}
+
+/*
+getType converts the given OpenAPI v3 `schema` to a ObjectTypeSpec and adds
+it to the `types` map under the given `name`. Recursively converts and adds all
+nested schemas as well.
+*/
```
```suggestion
// getType converts the given OpenAPI v3 `schema` to a ObjectTypeSpec and adds
// it to the `types` map under the given `name`. Recursively converts and adds all
// nested schemas as well.
```

lblackstone commented on crds/main.go:
```diff
@@ -0,0 +1,374 @@
+	properties, _, _ := unstruct.NestedMap(schema, "properties")
+	required, _, _ := unstruct.NestedStringSlice(schema, "required")
+
+	propertySpecs := make(map[string]pschema.PropertySpec)
```
FYI, you can write this like `map[string]pschema.PropertySpec{}` to save a little typing over explicitly calling `make`. Mostly a matter of style.

lblackstone commented on crds/main.go:
```diff
@@ -0,0 +1,374 @@
+	}
+}
+
+// cleanMetadataName replaces all comma-seperated instances of "pulumi" in the
```
Looks like this is "dot-separated" rather than "comma-separated"

lblackstone commented on crds/main.go:
```diff
@@ -0,0 +1,374 @@
+// schema. If combineRequired is true, then required properties are also
+// combined into the returned schema; otherwise all properties in the returned
+// schema are optional.
+func combineSchemas(combineRequired bool, schemas ...map[string]interface{}) map[string]interface{} {
```
It would be good to add a unit test to make sure this is working as expected. This logic is hard to verify by inspection.

lblackstone commented on crds/main.go:
```diff
@@ -0,0 +1,374 @@
+// Copyright 2016-2020, Pulumi Corporation.
```
Let's move these files to the `provider/cmd` directory. https://github.com/pulumi/pulumi-kubernetes/tree/master/provider/cmd

albert-zhong commented:
Just finished implementing (almost) all of Justin and Levi's suggestions. Only thing left to do is create end-to-end testing for verifying if a YAML resource definition produces the correct Pulumi code, which I will do on Monday.

lblackstone commented on crds/go.mod:
```diff
@@ -0,0 +1,16 @@
+module albertzhong.com/crds
```
Are you still using the `crds` folder?

albert-zhong replied:
I should've deleted the `crds` folder, my bad

lblackstone commented on crds/go.mod:
```diff
@@ -0,0 +1,16 @@
+	k8s.io/apimachinery v0.18.0
+)
+
+replace github.com/pulumi/pulumi/pkg/v2 => ../../pulumi/pkg
```
(Assuming this file is still in use) This should be replaced with a pinned version of the pulumi/pulumi dep.
